Feature - Vitest test suite for utility functions

barrymun opened the pull request:
# Description

- Adding the `vitest` package.
- Writing a sample test for the `utils/slugify.ts` utility function.
- Updating the various ignore lists so we don't lint or spell check the test output files/directories.
- Update the `slugify` function so that unnecessary escape chars are removed (function tested before and after the changes using the tests).
- Update `pre-commit` scripts (including the GitHub workflow script) to include this new test command.

## Type of Change

<!-- What kind of change does this pull request introduce? (Check all that apply) -->

- [ ] ✨ New snippet
- [ ] 🛠 Improvement to an existing snippet
- [ ] 🐞 Bug fix
- [ ] 📖 Documentation update
- [x] 🔧 Other (please describe): Feature - Adding a new test suite for utility functions. Component testing suite to come later.

## Checklist

<!--  Before submitting, ensure your pull request meets these requirements: -->

- [x] I have tested my code and verified it works as expected.
- [x] My code follows the style and contribution guidelines of this project.
- [x] Comments are added where necessary for clarity.
- [ ] Documentation has been updated (if applicable).
- [x] There are no new warnings or errors from my changes.

## Related Issues

<!-- Link any relevant issues (use #issue-number syntax). If not, leave it empty -->

Closes #135 

## Additional Context

<!-- Add any extra details, questions, or considerations here. -->

## Screenshots (Optional)

<!-- If your changes affect visuals, please include screenshots. -->

<details> 
<summary>Click to view screenshots</summary> 

<!-- Add your screenshots here --> 

</details>

Mathys-Gasnier reviewed (changes requested):
Looks good to me, but i would prefer for tests to be in the `/tests` folder instead of mixed with the source code

barrymun commented:
> Looks good to me, but i would prefer for tests to be in the `/tests` folder instead of mixed with the source code

Thanks for reviewing. Personally I prefer to keep all files together (this could be the .ts|.tsx, the .test.ts|.tsx, the .css|.scss, the .stories, the .md (if applicable), etc.) within the same directory as (for me) it makes it much easier to find all associated files. Placing all test files within the `/tests` directory is fine too, but if the project gets bigger it may become more difficult to find tests directly associated to a given file, especially if file names are similar but within different parent directories.

It is a matter of preference though, so let me know your thoughts. Thanks

Mathys-Gasnier commented:
IMO the repo shouldn't grow that much bigger, it's still a small project, appart from the snippet quantity

barrymun commented:
@Mathys-Gasnier In that case we'll keep it simple. All test files are now contained within the `tests` directory. Thanks

Mathys-Gasnier commented:
Btw could you add more tests to your PR, so we don't merge just to merge one test and "forget" about it ?
Feel free to ask if you dont know what you could add tests for

barrymun commented:
I'm not sure what you mean - `slugify` was the only util function that I could see that could benefit from tests for the moment. Of course component tests would also be useful, but I was going to add this in a separate feature so that the work was split.

As for forgetting about tests - I think most contributors would also want to test a util function when added or modified so I do not think this will be a problem in the long run. Even if this occurs when drafting a PR the reviewers would catch it. The first line of the checklist also requests that contributors test their code; perhaps the language here could be extended to encompass the addition of unit tests where applicable?

Mathys-Gasnier commented:
There is the utils root folder, with the scripts to check/consolidate snippets, but one of the file `snippetParser.js` is a util, that could benefit from tests

barrymun commented:
Let me see what I can do.

barrymun commented:
@Mathys-Gasnier I've made a lot of changes to try and expand the test suite. Firstly, I've added the [tsx](https://tsx.is/) package and changed the scripts within the `/utils` directory so that they are now typescript files. tsx will then be used to run these files instead of node. This will ensure additional type safety, and it allows us to move some of the utility functions into the `src` directory where they can be re-used. I've also added more unit tests for these functions. I've also updated the references in the workflow scripts so that they will use tsx instead of node as well. This should be tested to make sure they are on working GitHub. I would also recommend you re-test the scripts yourself to make sure I haven't altered the functionality in an unexpected way.

barrymun commented on public/consolidated/javascript.json:
```diff
@@ -85,10 +85,8 @@
                 "description": "Converts RGB color values to hexadecimal color code.",
                 "author": "jjcantu",
                 "tags": [
-                    "javascript",
```
After running the `snippets:consolidate` script some of these tags were removed - is this desirable? If so, should this script entry also be included in the husky pre-commit hooks?

Mathys-Gasnier commented on utils/snippetParser.ts:
Seeing the repetition of `errored=true; raise(...);`
Could you put them inside their own little function to avoid a lot of repetition

barrymun replied:
I've reworked this part of the script to avoid duplicated [here](https://github.com/dostonnabotov/quicksnip/pull/137/commits/ccc4cc0a3675dfc562d2b6fabd8f4aef1d7e9ffd).

Mathys-Gasnier commented on src/utils/objectUtils.ts:
This could be a cool snippet BTW

barrymun replied:
Might be something worth adding :)

Mathys-Gasnier commented on .github/workflows/check-snippets.yml:
```diff
@@ -25,7 +25,7 @@ jobs:
         uses: int128/comment-action@v1
         with:
-          run: node utils/checkSnippetFormatting.js # Run the script located in the utils/ folder
+          run: tsx utils/checkSnippetFormatting.ts # Run the script located in the utils/ folder
```
This will not work as this workflow doesn't install deps

barrymun replied:
Thanks, I've added the additional step to install the deps.

Mathys-Gasnier commented on .github/workflows/check-snippets.yml:
```diff
@@ -25,7 +25,7 @@ jobs:
       - name: Check if snippets are formated correctly
         uses: int128/comment-action@v1
```
Btw, i was the one who added the comment-action, to try and send the error in the PR, but for some reason it fails, you got any idea why ?

barrymun replied:
I'm not sure, but if I were to hazard a guess I would say that the error code triggered by the script is not recognised by the workflow as an error.

Mathys-Gasnier replied:
I will take a look at it in a later PR, thanks for your opinion on that

Mathys-Gasnier reviewed (changes requested):
Impressive work, didn't knew `tsx` was a thing, I always use `ts-node`, some little changes, but overall really good work

Mathys-Gasnier commented:
Oh and forgot, why did changing to TS change the consolidated snippets ? it's just a change of order of the fields, but i'm not a big fan that it changes that

Mathys-Gasnier commented on utils/snippetParser.ts:
```diff
@@ -6,15 +6,23 @@ import { isCorrectType } from "../src/utils/objectUtils";
+
+let errored = false;
+
+function parseSnippet({
```
I don't like at all the fact that the arguments are passed as an object... It makes it harder to use for no real reason

barrymun replied:
Again, this is a personal preference thing :) The reason I do it (for more than 2 args in my case) is because I find that it can be very easy to mix up the order of the args passed, but an object structure makes it more difficult to do this, as you have to deliberately pass in an arg in an incorrect place.

I will revert it, however.

barrymun commented:
> Oh and forgot, why did changing to TS change the consolidated snippets ? it's just a change of order of the fields, but i'm not a big fan that it changes that

I reverted those as I did not mean for that to happen.

barrymun commented:
Thanks, I think the `tsx` package is very cool. I personally prefer using it over `ts-node` as I've had config issues with the latter in the past.

Mathys-Gasnier commented:
Once you are done, could you ping me on discord ? So i can clone your branch and test the scripts are working correctly

Mathys-Gasnier reviewed (approved):
Really good work, let's wait on another review and I think we are good to merge

psychlone77 reviewed (approved):
Looks good. 👍

barrymun commented:
Updated and ready for merge. I have tested the scripts locally but it might be worth checking these again as the scripts have changed since they were modified.

Mathys-Gasnier commented on public/consolidated/javascript.json:
You still have this consolidated file that shouldn't be here

barrymun replied:
Removed

Mathys-Gasnier commented on utils/snippetParser.ts:
```diff
@@ -0,0 +1,192 @@
+interface ParseLanguageResponse {
+  name: string;
+  icon: string;
+  categories: Array<{
```
Just wanted to know, why are you using `Array<T>` instead of `T[]` ?

barrymun replied:
I use it personally for custom types or for nested/more complex types/interfaces as I find it more readable, but both are valid syntax. `Array<T>` is just the generic syntax.

barrymun replied:
Would you like me to change this to not use the generic `Array` syntax?

barrymun replied:
Changes made.

Mathys-Gasnier replied:
I prefer it, I know it's less readable, but most people are used to it, and using `Array` might confuse new contributors

barrymun replied:
Thanks, do you think this one is ready for merge if you're happy with the refactored scripts?
